morph.go

```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
+		}
+
+		diff := make([]*models.Migration, 0, len(downMigrations))
+		for i := 0; i < len(sortedMigrations); i++ {
+			diff = append(diff, downMigrations[sortedMigrations[i].Name])
+		}
+
+		return diff, nil
+	}
+
+	pendingMigrations, err := computePendingMigrations(appliedMigrations, m.source.Migrations())
+	if err != nil {
+		return nil, err
+	}
+
+	var diff []*models.Migration
+	for _, migration := range sortMigrations(pendingMigrations) {
+		if migration.Direction != models.Up {
+			continue
+		}
+		diff = append(diff, migration)
+	}
+
+	return diff, nil
+}
+
+func (m *Morph) GetOppositeMigrations(migrations []*models.Migration) ([]*models.Migration, error) {
+	var direction models.Direction
+	migrationsMap := make(map[string]*models.Migration)
+	for _, migration := range migrations {
+		if direction == "" {
+			direction = migration.Direction
+		}
+		// check if the migrations has the same direction
+		if direction != migration.Direction {
+			return nil, errors.New("migrations have different directions")
+		}
+
+		migrationsMap[migration.Name] = migration
+	}
+
+	rollbackMigrations := make([]*models.Migration, 0, len(migrations))
+	availableMigrations := m.source.Migrations()
+	for _, migration := range availableMigrations {
+		// skip if we have the same direction for the migration
+		// we are looking for opposite direction
+		if migration.Direction == direction {
+			continue
+		}
+
+		// we don't have the migration in the map
+		// so we can't rollback it
+		_, ok := migrationsMap[migration.Name]
+		if !ok {
+			continue
+		}
+
+		rollbackMigrations = append(rollbackMigrations, migration)
+	}
+
+	if len(migrations) != len(rollbackMigrations) {
+		return nil, errors.New("not all migrations have opposite migrations")
+	}
+
+	return rollbackMigrations, nil
+}
+
+// GeneratePlan returns the plan to apply these migrations and also includes
+// the safe rollback steps for the given migrations.
+func (m *Morph) GeneratePlan(migrations []*models.Migration) (*models.Plan, error) {
+	rollbackMigrations, err := m.GetOppositeMigrations(migrations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get opposite migrations: %w", err)
+	}
+
+	plan := models.NewPlan(migrations, rollbackMigrations)
+
+	return plan, nil
+}
+
+func (m *Morph) ApplyPlan(plan *models.Plan) error {
+	if err := plan.Validate(); err != nil {
+		return fmt.Errorf("invalid plan: %w", err)
+	}
+
+	revertMigrations := make([]*models.Migration, 0, len(plan.RevertMigrations))
+
+	for i := range plan.Migrations {
+		// add to the revert queue
+		for _, migration := range plan.RevertMigrations {
+			if migration.Name == plan.Migrations[i].Name {
+				revertMigrations = append(revertMigrations, migration)
+				break
+			}
+		}
+
+		err := m.apply(plan.Migrations[i], true)
+		if err != nil {
+			// log the fail step
+			m.config.Logger.Printf("migration %s failed, starting rollback", plan.Migrations[i].Name)
+
+			for j := len(revertMigrations) - 1; j >= 0; j-- {
```

Comment:
Any reason we are starting from the very end, if we know that `i` might be < `len(revertMigrations)`?

This is what I was thinking:

if we have migration 1, 2, 3. Then 1 passed, 2 failed, we will just run revert migrations from 2,1. In your case, we are running reverse migrations for 3,2,1.

Reply:
Since we queued the revert migrations, we are reading it from the last to treat it as a stack. In my case it will work like this:
- add revert migration_1 to the queue
- apply migration 1
- no error continue
- add revert migration_2 to the queue
- apply migration 2
- error ->
  - apply revert migration_2 (since we are reading in reverse order)
  - apply revert migration_1
- return error

Reply:
Ah okay, I read the code incorrectly. So I was thinking of just using the index till where the forward loop ran, and then just use that index to traverse the reverseMigrations slices backwards. Then you don't need to create a stack and pop them. This felt a bit simpler mentally to me. But I'll leave it up to you.

Reply:
That's indeed much simpler, but it leans on the *right* order of the plan. We are now ensuring the order is right. That's why added the logic.